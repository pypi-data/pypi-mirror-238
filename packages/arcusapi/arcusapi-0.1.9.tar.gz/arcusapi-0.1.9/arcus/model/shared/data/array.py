# Copyright [2023] [Arcus Inc.]

# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at

#     http://www.apache.org/licenses/LICENSE-2.0

# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


from typing import List, Optional, Union

import numpy as np

from arcus.model.shared.data.external_feature_type import ExternalFeatureType

try:
    import torch

    ARRAY_TYPE = Union[np.ndarray, torch.Tensor]
except ImportError:
    ARRAY_TYPE = np.ndarray


class ExternalDataArray:
    """
    Represents an array that is obtained from the Arcus data platform.
    Args:
        external_data_values: The values of the data that is obtained from
            the Arcus data platform.
        feature_type: The type of the external data (raw or embeddings).
    """

    def __init__(
        self,
        external_data_values: ARRAY_TYPE,
        feature_type: Optional[ExternalFeatureType] = ExternalFeatureType.RAW,
    ):
        """
        Represents a tensor that is obtained from the Arcus data platform.
        """
        self.external_data_values = external_data_values
        self.feature_type = feature_type

    def __len__(self):
        return len(self.external_data_values)

    @property
    def shape(self):
        return self.external_data_values.shape

    def __getitem__(self, index: Union[int, List[int], slice, ARRAY_TYPE]):
        return ExternalDataArray(
            self.external_data_values[index],
            self.feature_type,
        )

    def is_raw(self) -> bool:
        return self.feature_type is ExternalFeatureType.RAW

    def is_embedding(self) -> bool:
        return self.feature_type is ExternalFeatureType.EMBEDDING

    def get_data(self) -> np.ndarray:
        return self.external_data_values
