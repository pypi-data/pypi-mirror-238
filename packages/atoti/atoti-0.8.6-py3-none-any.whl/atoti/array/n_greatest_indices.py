from __future__ import annotations

from .._measure_convertible import MeasureConvertible, NonConstantMeasureConvertible
from .._measure_description import MeasureDescription, convert_to_measure_description
from .._measures.calculated_measure import CalculatedMeasure, Operator
from ._utils import check_array_type, validate_n_argument


def n_greatest_indices(
    measure: NonConstantMeasureConvertible, /, n: MeasureConvertible
) -> MeasureDescription:
    """Return an array measure containing the indices of the *n* greatest elements of the passed array measure.

    The indices in the returned array are sorted, so the first index corresponds to the greatest value and the last index to to the n-th greatest value.

    Example:
        >>> pnl_table = session.read_csv(
        ...     f"{RESOURCES}/pnl.csv",
        ...     array_separator=";",
        ...     keys=["Continent", "Country"],
        ...     table_name="PnL",
        ... )
        >>> cube = session.create_cube(pnl_table)
        >>> l, m = cube.levels, cube.measures
        >>> m["Top 3 indices"] = tt.array.n_greatest_indices(m["PnL.SUM"], n=3)
        >>> cube.query(m["PnL.SUM"], m["Top 3 indices"])
                                  PnL.SUM         Top 3 indices
        0  doubleVector[10]{-20.163, ...}  intVector[3]{6, ...}

    """
    validate_n_argument(n)
    check_array_type(measure)
    return CalculatedMeasure(
        Operator(
            "n_greatest_indices",
            [convert_to_measure_description(arg) for arg in [measure, n]],
        )
    )
