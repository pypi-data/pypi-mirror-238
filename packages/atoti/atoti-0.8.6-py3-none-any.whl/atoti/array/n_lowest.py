from __future__ import annotations

from .._measure_convertible import MeasureConvertible, NonConstantMeasureConvertible
from .._measure_description import MeasureDescription, convert_to_measure_description
from .._measures.calculated_measure import CalculatedMeasure, Operator
from ._utils import check_array_type, validate_n_argument


def n_lowest(
    measure: NonConstantMeasureConvertible, /, n: MeasureConvertible
) -> MeasureDescription:
    """Return an array measure containing the *n* lowest elements of the passed array measure.

    The values in the returned array are not sorted, use :func:`atoti.array.sort` to sort them.

    Example:
        >>> pnl_table = session.read_csv(
        ...     f"{RESOURCES}/pnl.csv",
        ...     array_separator=";",
        ...     keys=["Continent", "Country"],
        ...     table_name="PnL",
        ... )
        >>> cube = session.create_cube(pnl_table)
        >>> l, m = cube.levels, cube.measures
        >>> m["Bottom 5"] = tt.array.n_lowest(m["PnL.SUM"], n=5)
        >>> cube.query(m["PnL.SUM"], m["Bottom 5"])
                                  PnL.SUM                       Bottom 5
        0  doubleVector[10]{-20.163, ...}  doubleVector[5]{-20.163, ...}

    """
    validate_n_argument(n)
    check_array_type(measure)
    return CalculatedMeasure(
        Operator(
            "n_lowest", [convert_to_measure_description(arg) for arg in [measure, n]]
        )
    )
