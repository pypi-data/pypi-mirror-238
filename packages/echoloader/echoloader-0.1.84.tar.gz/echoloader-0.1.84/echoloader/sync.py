import base64
import datetime
import io
import itertools
import logging
import ssl
import threading
import time
from collections import defaultdict
from functools import cached_property

import requests
from dateutil import parser
from pathy import Pathy
from pydicom import dcmread
from pynetdicom import AE

from echoloader.hl7 import Hl7
from echoloader.login import unpack

logger = logging.getLogger('echolog')
DEFAULT_AE_TITLE = "Us2.ai"


def server_context(ca, cert, key):
    context = ssl.create_default_context(ssl.Purpose.CLIENT_AUTH)
    context.verify_mode = ssl.CERT_REQUIRED
    context.load_cert_chain(certfile=cert, keyfile=key)
    context.load_verify_locations(cafile=ca)
    # Only TLS <= 1.2 is supported, make sure we always use this
    context.minimum_version = context.maximum_version = ssl.TLSVersion.TLSv1_2

    return context


def client_context(ca, cert, key):
    context = ssl.create_default_context(ssl.Purpose.SERVER_AUTH, cafile=ca)
    context.verify_mode = ssl.CERT_REQUIRED
    context.load_cert_chain(certfile=cert, keyfile=key)
    context.check_hostname = False

    return context


class PacsConnection:
    def __init__(self, details):
        self.store_on_filesystem = True
        self.host = self.port = self.remote_ae_title = self.local_ae_title = self.cert = self.key = None
        self.details = details
        parts = details.split(':')
        if len(parts) > 2:
            self.store_on_filesystem = False
            self.host, port, self.remote_ae_title = parts[:3]
            self.local_ae_title = parts[3] if len(parts) > 3 else DEFAULT_AE_TITLE
            self.ca, self.cert, self.key = parts[4:7] if len(parts) > 5 else (None, None, None)
            self.port = int(port)

    def store(self, ds, called_ae):
        if self.store_on_filesystem:
            dst = Pathy.fluid(self.details) / ds.PatientID / f"{ds.SOPInstanceUID}.dcm"
            dst.parent.mkdir(exist_ok=True, parents=True)
            ds.save_as(str(dst), write_like_original=False)
            return
        ae = AE(ae_title=self.local_ae_title)
        ae.add_requested_context(ds.SOPClassUID, ds.file_meta.TransferSyntaxUID)

        remote_ae = called_ae or self.remote_ae_title
        assoc = ae.associate(self.host, self.port, ae_title=remote_ae,
                             tls_args=(client_context(self.ca, self.cert, self.key), None) if self.cert else None)

        if not assoc.is_established:
            raise ConnectionError('Association rejected, aborted or never connected')
        # Use the C-STORE service to send the dataset
        # returns the response status as a pydicom Dataset
        try:
            # force treat context as supporting the SCP role
            for cx in assoc.accepted_contexts:
                cx._as_scp = True

            status = assoc.send_c_store(ds)

            # Check the status of the storage request
            if status:
                # If the storage request succeeded this will be 0x0000
                logger.debug(f'C-STORE request status: 0x{status.Status:04x}')
            else:
                raise ValueError('Connection timed out, was aborted or received invalid response')
        finally:
            # Release the association
            assoc.release()

    def __str__(self):
        return self.details


class Sync(threading.Thread):
    def __init__(self, cmd, pool, *vargs, **kwargs):
        super().__init__(*vargs, **kwargs)
        self.args = cmd
        self.by_measurement = cmd.sync_by_measurement
        self.connections = cmd.sync
        self.auth = cmd.auth
        self.api_url = self.auth.api_url
        self.uploader = self.auth.username
        self.killed = False
        self.params = {'v': cmd.v}
        self.sync_from = eval(cmd.sync_from).replace(tzinfo=datetime.timezone.utc)
        self.last_sync = {}
        self.modalities = cmd.sync_modalities
        self.poll = cmd.sync_poll
        self.sr_params = {}
        self.doc_params = {}
        self.pdf_params = {}
        self.pool = pool
        self.search_params = {k: v for e in cmd.sync_search for k, v in [e.split('=', 1)]}
        self.protocol = unpack(requests.get(
            f'{self.api_url}/sync/protocol', params=self.params, headers=self.auth.get_headers()))['current_protocol']
        if cmd.sync_url:
            self.sr_params['url'] = True
        if cmd.sync_main_findings:
            self.sr_params['main_findings'] = True
            self.doc_params['main_findings'] = True
            self.pdf_params['main_findings'] = True
        if cmd.sync_pdf_images:
            self.doc_params['image'] = True
            self.pdf_params['image'] = True
        if cmd.sync_designators:
            self.sr_params['designators'] = cmd.sync_designators
        if cmd.sync_mapping:
            self.sr_params['mapping'] = cmd.sync_mapping
        if cmd.sync_regulatory_status:
            self.sr_params['regulatory_status'] = True
        if cmd.sync_edited_status:
            self.sr_params['edited_status'] = True
        if cmd.sync_annotations:
            self.sr_params['annotations'] = True
        self.doc_params['dicom_encapsulated_pdf'] = True
        self.ps_params = {}
        self.sc_params = {}

    def latest_mod(self, sid):
        resp = unpack(requests.get(
            f"{self.api_url}/sync/modification/{sid}",
            params={**self.params, 'limit': 1},
            headers=self.auth.get_headers()))
        mods = resp['results'] if isinstance(resp, dict) else resp
        return mods[-1]

    def stop(self):
        self.killed = True

    def handle_study_sync_error(self, err, sid):
        logger.error(f'Failed to sync study {sid} due to {err}')

    def sync(self):
        filter_params = {
            **self.params,
            'uploader': self.uploader,
            'lastUpdatedFrom': max([self.sync_from, *self.last_sync.values()]),
            **self.search_params,
        }
        res = unpack(
            requests.get(f'{self.api_url}/study/search', params=filter_params, headers=self.auth.get_headers()), {})
        results = res.get('results', [])
        for study in results:  # All search results have been updated since we last checked -> sync everything
            sid = study['id']
            last_sync = self.last_sync.get(sid, self.sync_from)
            logger.info(f'Syncing {sid} for changes since {last_sync}')
            mod = self.latest_mod(sid)
            creation = parser.parse(mod['creation']).replace(tzinfo=datetime.timezone.utc)
            self.last_sync[sid] = creation
            sync = SyncWorker(self, study, last_sync)
            self.pool.apply_async(sync.sync_study, error_callback=lambda err: self.handle_study_sync_error(err, sid))

    def run(self) -> None:
        while not self.killed:
            try:
                self.sync()
            except Exception as exc:
                logger.error(f'Failed sync due to: {exc}')
            time.sleep(self.poll)


class SyncWorker:
    def __init__(self, worker: Sync, study, t):
        self.api_url = worker.api_url
        self.auth = worker.auth
        self.args = worker.args
        self.params = worker.params
        self.sr_params = worker.sr_params
        self.doc_params = worker.doc_params
        self.pdf_params = worker.pdf_params
        self.ps_params = worker.ps_params
        self.sc_params = worker.sc_params
        self.protocol = worker.protocol
        self.modalities = worker.modalities
        self.by_measurement = worker.by_measurement
        self.connections = worker.connections
        self.hl7_config = worker.auth.user.get('dicom_router_config', {}).get('hl7_config', {})
        self.study = study
        self.sid = study['id']
        self.t = t

    def sr(self):
        return requests.get(f"{self.api_url}/study/sr/{self.sid}", headers=self.auth.get_headers(),
                            params={**self.params, **self.sr_params})

    def ds(self):
        ds = defaultdict(dict)
        for mod in self.mods:
            if mod['model'] == 'dicom.dicom':
                pk = mod['obj_pk']
                ds[pk].update(mod['new_fields'])
                ds[pk]['last_update'] = parser.parse(mod['creation']).replace(tzinfo=datetime.timezone.utc)
                if mod['action'] == 'delete' and pk in ds:
                    del ds[pk]
        for k, d in ds.items():
            if d['last_update'] > self.t and not d.get('from_dicom_id') and d.get('output_path'):
                yield requests.get(f"{self.api_url}/dicom/ds/{k}", headers=self.auth.get_headers(),
                                   params={**self.params})

    def ps(self, ms):
        return requests.get(f"{self.api_url}/dicom/ps", headers=self.auth.get_headers(),
                            params={**self.params, **self.ps_params, 'measurements': ms})

    def sc(self, ms):
        return requests.get(f"{self.api_url}/dicom/sc", headers=self.auth.get_headers(),
                            params={**self.params, **self.sc_params, 'measurements': ms})

    def doc(self):
        return requests.get(f"{self.api_url}/study/pdf/{self.sid}", headers=self.auth.get_headers(),
                            params={**self.params, **self.doc_params})

    def pdf(self):
        return requests.get(f"{self.api_url}/study/pdf/{self.sid}", headers=self.auth.get_headers(),
                            params={**self.params, **self.pdf_params})

    @cached_property
    def mods(self):
        params = {**self.params, 'page': 1, 'page_size': 10_000}
        result = []
        count = 1
        while len(result) < count:
            mods = unpack(requests.get(
                f"{self.api_url}/sync/modification/{self.sid}", params=params, headers=self.auth.get_headers()))
            result.extend(mods['results'] if isinstance(mods, dict) else mods)
            count = mods['count'] if isinstance(mods, dict) else len(mods)
            params['page'] += 1
        return result

    def read_measurements(self):
        ms = defaultdict(dict)
        for mod in self.mods:
            if mod['model'] == 'measurement.measurements':
                pk = mod['obj_pk']
                ms[pk].update(mod['new_fields'])
                ms[pk]['last_update'] = parser.parse(mod['creation']).replace(tzinfo=datetime.timezone.utc)
                if mod['action'] == 'delete' and pk in ms:
                    del ms[pk]
        measurements = {}
        for m in ms.values():
            proto = self.protocol.get('measurements', {}).get(str(m.get('code_id')), {})
            if (proto.get('shouldDisplay')
                    and m.get('used')
                    and m.get('dicom_id')
                    and m.get('plot_obj')):
                measurements[m['code_id']] = {
                    "proto": proto,
                    "m_value": m,
                }
        return measurements

    def media(self):
        ms = defaultdict(dict)
        for mod in self.mods:
            if mod['model'] == 'measurement.measurements':
                pk = mod['obj_pk']
                ms[pk].update(mod['new_fields'])
                ms[pk]['last_update'] = parser.parse(mod['creation']).replace(tzinfo=datetime.timezone.utc)
                if mod['action'] == 'delete' and pk in ms:
                    del ms[pk]
        grouped = defaultdict(list)
        for m in ms.values():
            proto = self.protocol.get('measurements', {}).get(str(m.get('code_id')), {})
            if (proto.get('shouldDisplay')
                    and m['last_update'] > self.t
                    and m.get('used')
                    and m.get('dicom_id')
                    and m.get('plot_obj')):
                k = (m['dicom_id'], m['frame'], *([m['id']] if self.by_measurement else []))
                grouped[k].append(m['id'])
        for ms in grouped.values():
            if 'PS' in self.modalities:
                yield self.ps(ms)
            if 'SC' in self.modalities:
                yield self.sc(ms)

    def sync_hl7(self):
        measurements = {}
        report_pdf_encoded = None

        try:
            report_type = self.hl7_config.get('report_type', 'TEXT')
            if report_type != "PDF":
                measurements = self.read_measurements()
            if report_type != "TEXT":
                report_pdf_response = self.pdf()
                if report_pdf_response.status_code == 200:
                    report_pdf_encoded = base64.b64encode(report_pdf_response.content).decode("utf-8")
                else:
                    logger.error(f'Failed to fetch from {report_pdf_response.url} - {report_pdf_response.status_code}')

            hl7 = Hl7(self.hl7_config, "ORU_R01", "2.5")
            msg_control_id = hl7.generate(self.study, measurements, report_type, report_pdf_encoded)

            if msg_control_id:
                hl7.send()
            else:
                logger.warning(f'Failed to generate HL7 {msg_control_id}')
        except Exception as ex:
            logger.error(f'Failed to sync HL7 due to {ex}')

    def sync_study(self):
        for req in itertools.chain(
                self.media() if [m in self.modalities for m in ['PS', 'SC']] else [],
                self.ds() if 'DS' in self.modalities else [],
                [self.sr()] if 'SR' in self.modalities else [],
                [self.doc()] if 'DOC' in self.modalities else [],
        ):
            url = req.url
            try:
                bs = unpack(req)
            except Exception as exc:
                logger.error(f'Failed to fetch {url} due to {exc}')
                continue
            ds = dcmread(io.BytesIO(bs))
            for conn in self.connections:
                try:
                    called_ae = None
                    if self.args.customer_aet:
                        called_ae = self.study.get('customer')

                    conn.store(ds, called_ae)
                    logger.info(f'Synced {url} to {conn}')
                except Exception as exc:
                    logger.error(f'Failed to sync {url} due to {exc}')

        if self.hl7_config.get('enabled', False):
            self.sync_hl7()
