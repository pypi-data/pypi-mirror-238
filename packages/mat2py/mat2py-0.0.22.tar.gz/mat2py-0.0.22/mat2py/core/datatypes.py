# type: ignore

__all__ = [
    "iscom",
    "orderfields",
    "deal",
    "uint16",
    "typecast",
    "inferiorto",
    "int32",
    "function_handle",
    "functions",
    "isinterface",
    "findgroups",
    "double",
    "_class",
    "isenum",
    "str2func",
    "javaArray",
    "array2table",
    "cell2struct",
    "cast",
    "fieldnames",
    "single",
    "struct2cell",
    "getfield",
    "rmfield",
    "iscell",
    "logical",
    "isjava",
    "setfield",
    "javaMethod",
    "table2cell",
    "methods",
    "int8",
    "saveobj",
    "properties",
    "swapbytes",
    "substruct",
    "uint64",
    "superclasses",
    "uint8",
    "isnumeric",
    "isobject",
    "istable",
    "superiorfloat",
    "isfield",
    "javaObject",
    "islogical",
    "cell2mat",
    "ismethod",
    "javaObjectEDT",
    "structfun",
    "cellfun",
    "int64",
    "enumeration",
    "splitapply",
    "isinteger",
    "isstruct",
    "table2struct",
    "isa",
    "num2cell",
    "struct2table",
    "superiorto",
    "events",
    "mat2cell",
    "func2str",
    "cell",
    "methodsview",
    "loadobj",
    "celldisp",
    "cellplot",
    "table2array",
    "isprop",
    "int16",
    "cell2table",
    "iscategorical",
    "uint32",
    "arrayfun",
    "struct",
    "isfloat",
    "javaMethodEDT",
    "metaclass",
]

from mat2py.common.backends import numpy as np

from ._internal import M
from ._internal.cell import CellArray, cell
from ._internal.math_helper import mp_type_cast_decorators
from ._internal.struct import StructArray, struct


def iscom(*args):
    raise NotImplementedError("iscom")


def orderfields(*args):
    raise NotImplementedError("orderfields")


def deal(*args):
    raise NotImplementedError("deal")


def uint16(*args):
    raise NotImplementedError("uint16")


def typecast(*args):
    raise NotImplementedError("typecast")


def inferiorto(*args):
    raise NotImplementedError("inferiorto")


def int32(*args):
    raise NotImplementedError("int32")


def function_handle(*args):
    raise NotImplementedError("function_handle")


def functions(*args):
    raise NotImplementedError("functions")


def isinterface(*args):
    raise NotImplementedError("isinterface")


def findgroups(*args):
    raise NotImplementedError("findgroups")


@mp_type_cast_decorators()
def double(dtype):
    return np.complex128 if np.issubdtype(dtype, np.complex_) else np.float64


def _class(*args):
    raise NotImplementedError("_class")


def isenum(*args):
    raise NotImplementedError("isenum")


def str2func(*args):
    raise NotImplementedError("str2func")


def javaArray(*args):
    raise NotImplementedError("javaArray")


def array2table(*args):
    raise NotImplementedError("array2table")


def cell2struct(*args):
    raise NotImplementedError("cell2struct")


def cast(*args):
    raise NotImplementedError("cast")


from ._internal.struct import fieldnames


@mp_type_cast_decorators()
def single(dtype):
    return np.complex64 if np.issubdtype(dtype, np.complex_) else np.float32


def struct2cell(*args):
    raise NotImplementedError("struct2cell")


def getfield(*args):
    raise NotImplementedError("getfield")


def rmfield(*args):
    raise NotImplementedError("rmfield")


def iscell(*args):
    raise NotImplementedError("iscell")


def logical(*args):
    raise NotImplementedError("logical")


def isjava(*args):
    raise NotImplementedError("isjava")


def setfield(*args):
    raise NotImplementedError("setfield")


def javaMethod(*args):
    raise NotImplementedError("javaMethod")


def table2cell(*args):
    raise NotImplementedError("table2cell")


def methods(*args):
    raise NotImplementedError("methods")


def int8(*args):
    raise NotImplementedError("int8")


def saveobj(*args):
    raise NotImplementedError("saveobj")


def properties(*args):
    raise NotImplementedError("properties")


def swapbytes(*args):
    raise NotImplementedError("swapbytes")


def substruct(*args):
    raise NotImplementedError("substruct")


def uint64(*args):
    raise NotImplementedError("uint64")


def superclasses(*args):
    raise NotImplementedError("superclasses")


def uint8(*args):
    raise NotImplementedError("uint8")


def isnumeric(*args):
    raise NotImplementedError("isnumeric")


def isobject(*args):
    raise NotImplementedError("isobject")


def istable(*args):
    raise NotImplementedError("istable")


def superiorfloat(*args):
    raise NotImplementedError("superiorfloat")


def isfield(*args):
    raise NotImplementedError("isfield")


def javaObject(*args):
    raise NotImplementedError("javaObject")


def islogical(*args):
    raise NotImplementedError("islogical")


def cell2mat(*args):
    raise NotImplementedError("cell2mat")


def ismethod(*args):
    raise NotImplementedError("ismethod")


def javaObjectEDT(*args):
    raise NotImplementedError("javaObjectEDT")


def structfun(*args):
    raise NotImplementedError("structfun")


def cellfun(*args):
    raise NotImplementedError("cellfun")


def int64(*args):
    raise NotImplementedError("int64")


def enumeration(*args):
    raise NotImplementedError("enumeration")


def splitapply(*args):
    raise NotImplementedError("splitapply")


def isinteger(*args):
    raise NotImplementedError("isinteger")


def isstruct(*args):
    raise NotImplementedError("isstruct")


def table2struct(*args):
    raise NotImplementedError("table2struct")


def isa(*args):
    raise NotImplementedError("isa")


def num2cell(*args):
    raise NotImplementedError("num2cell")


def struct2table(*args):
    raise NotImplementedError("struct2table")


def superiorto(*args):
    raise NotImplementedError("superiorto")


def events(*args):
    raise NotImplementedError("events")


def mat2cell(*args):
    raise NotImplementedError("mat2cell")


def func2str(*args):
    raise NotImplementedError("func2str")


def methodsview(*args):
    raise NotImplementedError("methodsview")


def loadobj(*args):
    raise NotImplementedError("loadobj")


def celldisp(*args):
    raise NotImplementedError("celldisp")


def cellplot(*args):
    raise NotImplementedError("cellplot")


def table2array(*args):
    raise NotImplementedError("table2array")


def isprop(*args):
    raise NotImplementedError("isprop")


def int16(*args):
    raise NotImplementedError("int16")


def cell2table(*args):
    raise NotImplementedError("cell2table")


def iscategorical(*args):
    raise NotImplementedError("iscategorical")


def uint32(*args):
    raise NotImplementedError("uint32")


def arrayfun(*args):
    raise NotImplementedError("arrayfun")


def isfloat(*args):
    raise NotImplementedError("isfloat")


def javaMethodEDT(*args):
    raise NotImplementedError("javaMethodEDT")


def metaclass(*args):
    raise NotImplementedError("metaclass")
