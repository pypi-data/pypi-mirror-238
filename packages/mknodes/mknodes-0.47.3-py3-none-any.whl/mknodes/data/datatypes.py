from __future__ import annotations

from collections.abc import Callable
from dataclasses import Field
import types

from typing import Any, ClassVar, Literal, Protocol


AdmonitionTypeStr = Literal[
    "note",
    "abstract",
    "info",
    "tip",
    "success",
    "question",
    "warning",
    "failure",
    "danger",
    "bug",
    "example",
    "quote",
]

ClassifierStr = Literal[
    "Development Status",
    "Environment",
    "Framework",
    "Intended Audience",
    "License",
    "Natural Language",
    "Operating System",
    "Programming Language",
    "Topic",
    "Typing",
]

CLASSIFIERS: list[ClassifierStr] = [
    "Development Status",
    "Environment",
    "Framework",
    "Intended Audience",
    "License",
    "Natural Language",
    "Operating System",
    "Programming Language",
    "Topic",
    "Typing",
]

EXT_TO_PYGMENTS_STYLE = {
    ".py": "py",
    ".yml": "yaml",
    ".yaml": "yaml",
    ".toml": "toml",
    ".md": "md",
    ".jinja": "jinja",
}

MetadataTypeStr = (
    Literal[
        "classifiers",
        "keywords",
        "keywords_combined",
        "websites",
        "dependencies",
        "required_python",
        "installed_packages",
    ]
    | ClassifierStr
)


class DataclassInstance(Protocol):
    __dataclass_fields__: ClassVar[dict[str, Field[Any]]]


PageStatusStr = Literal["new", "deprecated", "encrypted"]

RGBColorType = tuple[int, int, int] | str
RGBAColorType = tuple[int, int, int, float] | str
ColorType = RGBColorType | RGBAColorType
HasCodeType = (
    types.ModuleType
    | type
    | types.MethodType
    | types.FunctionType
    | types.TracebackType
    | types.FrameType
    | types.CodeType
    | Callable[..., Any]
)
