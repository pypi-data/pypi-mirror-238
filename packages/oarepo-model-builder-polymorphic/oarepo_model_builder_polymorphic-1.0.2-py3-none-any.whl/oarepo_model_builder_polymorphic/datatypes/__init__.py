from .polymorphic import PolymorphicComponent, PolymorphicDataType

datatypes = [PolymorphicDataType]

components = [PolymorphicComponent]
