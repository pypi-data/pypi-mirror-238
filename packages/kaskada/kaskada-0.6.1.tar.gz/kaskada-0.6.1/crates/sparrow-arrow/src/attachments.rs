use arrow::datatypes::SchemaRef;
use arrow::record_batch::RecordBatch;

#[derive(Debug)]
pub struct RecordBatchAttachment {
    label: &'static str,
    batch: RecordBatch,
}

impl RecordBatchAttachment {
    pub fn new(label: &'static str, batch: &RecordBatch) -> Self {
        Self {
            label,
            batch: batch.clone(),
        }
    }
}

impl std::fmt::Display for RecordBatchAttachment {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let num_rows = self.batch.num_rows();
        if num_rows < 5 {
            let batch = batch_to_json(self.batch.clone())?;
            write!(f, "Record Batch: {}\n{batch}", self.label)
        } else {
            let prefix = batch_to_json(self.batch.slice(0, 2))?;
            let suffix = batch_to_json(self.batch.slice(num_rows - 3, 2))?;

            let remainder = num_rows - 4;
            write!(
                f,
                "Record Batch: {}\n{prefix}... {remainder} extra rows...\n{suffix}",
                self.label
            )
        }
    }
}

fn batch_to_json(batch: RecordBatch) -> Result<String, std::fmt::Error> {
    let mut json_string = Vec::new();
    let mut writer = arrow::json::LineDelimitedWriter::new(&mut json_string);

    writer.write_batches(&[&batch]).map_err(|e| {
        tracing::error!("Error formatting batch: {}", e);
        std::fmt::Error
    })?;
    writer.finish().map_err(|e| {
        tracing::error!("Error formatting batch: {}", e);
        std::fmt::Error
    })?;
    let json_string = String::from_utf8(json_string).map_err(|e| {
        tracing::error!("Error formatting batch: {}", e);
        std::fmt::Error
    })?;

    Ok(json_string)
}

#[derive(Debug)]

pub struct SchemaAttachment {
    label: &'static str,
    schema: SchemaRef,
}

impl SchemaAttachment {
    pub fn new(label: &'static str, schema: &SchemaRef) -> Self {
        Self {
            label,
            schema: schema.clone(),
        }
    }
}

impl std::fmt::Display for SchemaAttachment {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        writeln!(f, "Schema: {}\n", self.label)?;

        for (k, v) in self.schema.metadata() {
            writeln!(f, "  Metadata: '{k}' = '{v}'")?;
        }

        for field in self.schema.fields.iter() {
            let nullable = if field.is_nullable() {
                ""
            } else {
                "(non-nullable) "
            };
            write!(
                f,
                "  Field: '{}' {nullable}= {}",
                field.name(),
                field.data_type(),
            )?;
            if field.metadata().is_empty() {
                writeln!(f)?;
            } else {
                writeln!(f, " {:?}", field.metadata())?;
            }
        }

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use std::sync::Arc;

    use super::*;
    use arrow::array::{Array, ArrayRef, Int32Array, UInt64Array};
    use arrow::datatypes::{DataType, Field, Schema};

    #[test]
    fn test_batch_format() {
        let a: ArrayRef = Arc::new(Int32Array::from(vec![0, 1, 2]));
        let b: ArrayRef = Arc::new(UInt64Array::from(vec![None, Some(1), Some(8)]));

        let schema = Schema::new(vec![
            Field::new("a", a.data_type().clone(), false),
            Field::new("b", b.data_type().clone(), true),
        ]);
        let schema = Arc::new(schema);
        let batch = RecordBatch::try_new(schema, vec![a, b]).unwrap();

        let annotation = RecordBatchAttachment::new("test batch", &batch);
        insta::assert_display_snapshot!(annotation)
    }

    #[test]
    fn test_long_batch_format() {
        let a: ArrayRef = Arc::new(Int32Array::from(vec![0, 1, 2, 3, 4, 5, 6, 7, 8, 9]));
        let b: ArrayRef = Arc::new(UInt64Array::from(vec![
            None,
            Some(1),
            Some(8),
            None,
            None,
            None,
            Some(2),
            Some(8),
            Some(9),
            Some(8),
        ]));

        let schema = Schema::new(vec![
            Field::new("a", a.data_type().clone(), false),
            Field::new("b", b.data_type().clone(), true),
        ]);
        let schema = Arc::new(schema);
        let batch = RecordBatch::try_new(schema, vec![a, b]).unwrap();

        let annotation = RecordBatchAttachment::new("test batch", &batch);
        insta::assert_display_snapshot!(annotation)
    }

    #[test]
    fn test_schema_format() {
        let schema = Schema::new(vec![
            Field::new("a", DataType::Int32, false),
            Field::new("b", DataType::Utf8, true),
        ]);
        let schema = Arc::new(schema);

        let annotation = SchemaAttachment::new("test schema", &schema);
        insta::assert_display_snapshot!(annotation)
    }
}
