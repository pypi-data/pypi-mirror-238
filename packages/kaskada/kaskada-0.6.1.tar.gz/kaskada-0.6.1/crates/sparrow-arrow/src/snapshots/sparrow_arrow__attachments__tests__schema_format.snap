---
source: crates/sparrow-arrow/src/attachments.rs
expression: annotation
---
Schema: test schema

  Field: 'a' (non-nullable) = Int32
  Field: 'b' = Utf8

